flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/FlinkKubeClientFactory.java

```diff
@@ -87,12 +89,23 @@ public FlinkKubeClient fromConfiguration(Configuration flinkConfig, Executor ioE
         LOG.debug("Setting namespace of Kubernetes client to {}", namespace);
         config.setNamespace(namespace);
 
-        final NamespacedKubernetesClient client = new DefaultKubernetesClient(config);
+        final Tuple2<String, String> cacheKey = Tuple2.of(config.getMasterUrl(), namespace);
+        if (CACHE.containsKey(cacheKey)) {
+            return CACHE.get(cacheKey);
+        }
 
-        return new Fabric8FlinkKubeClient(flinkConfig, client, () -> ioExecutor);
+        final NamespacedKubernetesClient client = new DefaultKubernetesClient(config);
+        final int poolSize =
+                flinkConfig.get(KubernetesConfigOptions.KUBERNETES_CLIENT_IO_EXECUTOR_POOL_SIZE);
+        final FlinkKubeClient flinkKubeClient =
+                new Fabric8FlinkKubeClient(
+                        flinkConfig, client, () -> createThreadPoolForAsyncIO(poolSize));
+        CACHE.put(cacheKey, flinkKubeClient);
+        return flinkKubeClient;
     }
 
-    private static Executor createThreadPoolForAsyncIO() {
-        return Executors.newFixedThreadPool(2, new ExecutorThreadFactory("FlinkKubeClient-IO"));
+    private static ExecutorService createThreadPoolForAsyncIO(int poolSize) {
+        return Executors.newFixedThreadPool(
+                poolSize, new ExecutorThreadFactory("flink-kubeclient-io"));
```

Comment:
If we create multiple `FlinkKubeClients`, then we might wanna give their thread pools different names.